app/src/main/scala/org/alephium/explorer/service/BlockFlowSyncService.scala

```diff
@@ -274,6 +274,17 @@ case object BlockFlowSyncService extends StrictLogging {
     }
   }
 
+  private def insertWithEvents(blockWithEvents: BlockEntityWithEvents)(
+      implicit ec: ExecutionContext,
+      dc: DatabaseConfig[PostgresProfile],
+      blockFlowClient: BlockFlowClient,
+      cache: BlockCache,
+      groupSetting: GroupSetting): Future[Unit] = {
+    for {
+      _ <- insert(blockWithEvents.block)
+      _ <- BlockDao.insertEvents(blockWithEvents.events)
+    } yield ()
+  }
```

Comment:
If `insert` succeeds but `insertEvents` fails (or explorer gets stopped), then on retry/reboot, this not being a transaction wouldn't lead to an inconsistent database state?

Reply:
totally! I'll try to fix that

Reply:
done